From d445d2084942cf64468f65a105b2a419c8f2df5f Mon Sep 17 00:00:00 2001
From: Ecco Park <eccopark@broadcom.com>
Date: Thu, 7 Nov 2013 09:29:47 +0900
Subject: [PATCH 212/338] net: wireless: bcmdhd: add debug code to print
 packet statistics

	I add the debug code to print packets statistics after resuming
	host.
	This will print only once after resume to see what packet cause
	to wake up host.

	Bug : 11374623
Signed-off-by: Ecco Park <eccopark@broadcom.com>
Signed-off-by: Simarpreet Singh <simar@linux.com>
---
 drivers/net/wireless/bcmdhd/Makefile             |    2 +-
 drivers/net/wireless/bcmdhd/bcmsdh_sdmmc_linux.c |    3 ++-
 drivers/net/wireless/bcmdhd/dhd_linux.c          |   31 +++++++++++++++++-----
 3 files changed, 28 insertions(+), 8 deletions(-)

diff --git a/drivers/net/wireless/bcmdhd/Makefile b/drivers/net/wireless/bcmdhd/Makefile
index dc070e3..b393e56 100644
--- a/drivers/net/wireless/bcmdhd/Makefile
+++ b/drivers/net/wireless/bcmdhd/Makefile
@@ -19,7 +19,7 @@ DHDCFLAGS = -Wall -Wstrict-prototypes -Dlinux -DBCMDRIVER               \
 	-DESCAN_RESULT_PATCH -DSDIO_CRC_ERROR_FIX \
 	-DDHD_DONOT_FORWARD_BCMEVENT_AS_NETWORK_PKT -DRXFRAME_THREAD	        \
 	-DSUPPORT_PM2_ONLY -DWLTDLS     \
-	-DMIRACAST_AMPDU_SIZE=8		\
+	-DMIRACAST_AMPDU_SIZE=8 -DDHD_RX_DUMP	\
 	-Idrivers/net/wireless/bcmdhd	-Idrivers/net/wireless/bcmdhd/include   \
 	-Idrivers/net/wireless/bcmdhd/common/include
 
diff --git a/drivers/net/wireless/bcmdhd/bcmsdh_sdmmc_linux.c b/drivers/net/wireless/bcmdhd/bcmsdh_sdmmc_linux.c
index 17e4177..2888202 100644
--- a/drivers/net/wireless/bcmdhd/bcmsdh_sdmmc_linux.c
+++ b/drivers/net/wireless/bcmdhd/bcmsdh_sdmmc_linux.c
@@ -103,7 +103,7 @@ PBCMSDH_SDMMC_INSTANCE gInstance;
 extern int bcmsdh_probe(struct device *dev);
 extern int bcmsdh_remove(struct device *dev);
 extern volatile bool dhd_mmc_suspend;
-
+extern bool bsuspend;
 static int bcmsdh_sdmmc_probe(struct sdio_func *func,
                               const struct sdio_device_id *id)
 {
@@ -219,6 +219,7 @@ static int bcmsdh_sdmmc_suspend(struct device *pdev)
 	bcmsdh_oob_intr_set(0);
 #endif 
 	dhd_mmc_suspend = TRUE;
+	bsuspend = TRUE;
 	smp_mb();
 
 	return 0;
diff --git a/drivers/net/wireless/bcmdhd/dhd_linux.c b/drivers/net/wireless/bcmdhd/dhd_linux.c
index 4ef7c70..f5c0813 100644
--- a/drivers/net/wireless/bcmdhd/dhd_linux.c
+++ b/drivers/net/wireless/bcmdhd/dhd_linux.c
@@ -138,6 +138,7 @@ static struct notifier_block dhd_notifier_ipv6 = {
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)) && defined(CONFIG_PM_SLEEP)
 #include <linux/suspend.h>
 volatile bool dhd_mmc_suspend = FALSE;
+bool bsuspend = FALSE;
 DECLARE_WAIT_QUEUE_HEAD(dhd_dpc_wait);
 #endif /* (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)) && defined(CONFIG_PM_SLEEP) */
 
@@ -1777,6 +1778,10 @@ dhd_rx_frame(dhd_pub_t *dhdp, int ifidx, void *pktbuf, int numpkt, uint8 chan)
 #endif /* DHD_RX_FULL_DUMP */
 	char *dump_data;
 	uint16 protocol;
+	int unicast = 0;
+	int broadcast = 0;
+	int multicast = 0;
+	int brcm = 0;
 #endif /* DHD_RX_DUMP */
 
 	DHD_TRACE(("%s: Enter\n", __FUNCTION__));
@@ -1837,7 +1842,7 @@ dhd_rx_frame(dhd_pub_t *dhdp, int ifidx, void *pktbuf, int numpkt, uint8 chan)
 #ifdef DHD_RX_DUMP
 		dump_data = skb->data;
 		protocol = (dump_data[12] << 8) | dump_data[13];
-		DHD_ERROR(("RX DUMP - %s\n", _get_packet_type_str(protocol)));
+		DHD_INFO(("RX DUMP - %s\n", _get_packet_type_str(protocol)));
 
 #ifdef DHD_RX_FULL_DUMP
 		if (protocol != ETHER_TYPE_BRCM) {
@@ -1852,24 +1857,29 @@ dhd_rx_frame(dhd_pub_t *dhdp, int ifidx, void *pktbuf, int numpkt, uint8 chan)
 
 		if (protocol != ETHER_TYPE_BRCM) {
 			if (dump_data[0] == 0xFF) {
-				DHD_ERROR(("%s: BROADCAST\n", __FUNCTION__));
-
+				DHD_INFO(("%s: BROADCAST\n", __FUNCTION__));
+				broadcast++;
 				if ((dump_data[12] == 8) &&
 					(dump_data[13] == 6)) {
-					DHD_ERROR(("%s: ARP %d\n",
+					DHD_INFO(("%s: ARP %d\n",
 						__FUNCTION__, dump_data[0x15]));
 				}
 			} else if (dump_data[0] & 1) {
-				DHD_ERROR(("%s: MULTICAST: " MACDBG "\n",
+				DHD_INFO(("%s: MULTICAST: " MACDBG "\n",
 					__FUNCTION__, MAC2STRDBG(dump_data)));
+				multicast++;
+			} else {
+				unicast++;
 			}
 
 			if (protocol == ETHER_TYPE_802_1X) {
-				DHD_ERROR(("ETHER_TYPE_802_1X: "
+				DHD_INFO(("ETHER_TYPE_802_1X: "
 					"ver %d, type %d, replay %d\n",
 					dump_data[14], dump_data[15],
 					dump_data[30]));
 			}
+		} else {
+			brcm++;
 		}
 
 #endif /* DHD_RX_DUMP */
@@ -1972,6 +1982,15 @@ dhd_rx_frame(dhd_pub_t *dhdp, int ifidx, void *pktbuf, int numpkt, uint8 chan)
 #endif
 	DHD_OS_WAKE_LOCK_RX_TIMEOUT_ENABLE(dhdp, tout_rx);
 	DHD_OS_WAKE_LOCK_CTRL_TIMEOUT_ENABLE(dhdp, tout_ctrl);
+	/* print once after waking up host */
+#ifdef DHD_RX_DUMP
+	if (bsuspend && tout_rx) {
+		/*flip the value */
+		bsuspend = FALSE;
+		DHD_ERROR(("DHD rx frame stats : unicast : %d, broadcast : %d, multicast : %d, brcm : %d",
+			unicast, broadcast, multicast, brcm));
+	}
+#endif
 }
 
 void
-- 
1.7.9.5

